### ISSUE TITLE
Router::is_graphql_request() returns false until the `init` action is fired.

### ISSUE THREAD

**Issue by kidunot89** [wp-graphql/wp-graphql#1021]
#### The Problem
I have WPGraphQL extension code that only wish to run during GraphQL requests before the `init` action is fired, however when I use `Router::is_graphql_request()` it always returns false.

#### The Expected Behavior
It should return if true on all GraphQL request at all steps in the request after the `Router` class is loaded.

#### Other Comments
The issue is the cause by the fact that the WP query vars aren't available until the `init` is fired. A alternate check should be used before the `do_action( `init` )` is called.

**Comment by esamattis** [comment]
I hit this too. My use case was to conditionally define acf rule for  page_template 

```php
if (!Router::is_graphql_request()) {
  // acf-codifier
  $rule_group->add_rule('page_template', '==', 'some-form');
}
```

I needed to do this because the rule was always false in graphql context and thus not showing the field in the schema. But this workaround did not work because `is_graphql_request()` returns false on the top level where the acf definition was.

**Comment by esamattis** [comment]
Also I originally contributed this method because I needed it for wp-graphql-polylang. But as it appears it does not work there neither because it needs it before the wp query vars are set.

My current hacky graphql context detection method is following: https://github.com/valu-digital/wp-graphql-polylang/blob/aac1be1657879d4590f2086c71794f3284315724/src/Loader.php#L108-L122

**Comment by jasonbahl** [comment]
@kidunot89 can you provide more insight on what you're trying to do conditionally within a GraphQL request _before_ the `init` hook? Some more context will help me understand the best way to approach here.

**Comment by jasonbahl** [comment]
If we're going to have conditionals like this, we have to also assume not everyone is using GraphQL via HTTP requests. 

The ACF example above, will only work when using ACF over HTTP, eh? Because the `Router::is_graphql_request()` is determined by the URI path. 

But if you wanted to use GraphQL within a PHP page template, for example, these wouldn't work because you'd be using the `graphql` or `do_graphql_request` method. 

We need to be careful about supporting these cases too.

**Comment by jasonbahl** [comment]
I wonder if we rename this to `Router::is_graphql_http_request()` so that at least there's a clear expectation that using this conditional will only validate HTTP requests, and not _all_ GraphQL requests. 

There are valid cases where you need to know if it's a GraphQL request coming over HTTP, but there are also the cases where we need to know if it's a GraphQL request, regardless of coming through the `/graphql` HTTP endpoint. 

I'm not sure the best way to solve this, if you need to know about it being a GraphQL request before `init` though, because adding a call to `graphql` as a resolver for a shortcode, or something, will execute later than `init` no matter what.

**Comment by esamattis** [comment]
> The ACF example above, will only work when using ACF over HTTP, eh?

Yeah. That's definitely a HTTP only solution. Not sure if `page_template` ACF rules can work at all from `graphql()` calls from PHP on templates because it may or may not have set the field which in turn makes the graphql schema dynamic based on the page it's being used on. But that discussion is offtopic for this issue.

> I wonder if we rename this to `Router::is_graphql_http_request()`

I'd be on board with that. When I earlier named it as just `is_graphql_request` I though the `Router` class would imply HTTP.

------

BTW on the subject of detecting graphql requests on templates isn't the usage of constant actually a really bad idea since it cannot be undefined?

There a some places where filters are active when the `GRAPHQL_REQUEST` is set to true.

For example in Config.php there are multiple

https://github.com/wp-graphql/wp-graphql/blob/4855753012815b13b4eb002e22d6ad407703ebfc/src/Data/Config.php#L250

https://github.com/wp-graphql/wp-graphql/blob/4855753012815b13b4eb002e22d6ad407703ebfc/src/Data/Config.php#L298

https://github.com/wp-graphql/wp-graphql/blob/4855753012815b13b4eb002e22d6ad407703ebfc/src/Data/Config.php#L223

https://github.com/wp-graphql/wp-graphql/blob/4855753012815b13b4eb002e22d6ad407703ebfc/src/Data/Config.php#L198

https://github.com/wp-graphql/wp-graphql/blob/4855753012815b13b4eb002e22d6ad407703ebfc/src/Data/Config.php#L179

https://github.com/wp-graphql/wp-graphql/blob/4855753012815b13b4eb002e22d6ad407703ebfc/src/Data/Config.php#L155

So when you do a graphql request from PHP using `graphql()` the constant `GRAPHQL_REQUEST` gets set to true and there for those filters are active even after the `graphql()`  call, right? That would be pretty bad if so.

**Comment by jasonbahl** [comment]
@esamattis ya, that's a good point. It seems like there's a legit need to be able to reset to false once a GraphQL request is completed, as there could be execution happening between multiple GraphQL requests 🤔

**Comment by esamattis** [comment]
So yeah `\WPGraphQL::is_graphql_http_request()` and `\WPGraphQL::is_graphql_request()` would be pretty clean API (and wp-graphql should start using the latter one internally).

**Comment by jasonbahl** [comment]
It still won't quite solve use cases where you need to know if it's a GraphQL request before `init`. 

I'm curious to learn more about that use case. Perhaps checking if it's a GraphQL HTTP Request will be enough, but then whatever case using that would not work in a non-HTTP GraphQL request. 

Curious if there are other approaches, or filters available for the "other thing" that's happening pre-init.

**Comment by esamattis** [comment]
Yeah, making `GRAPHQL_REQUEST` dyamic is a separate issue. Opened one here https://github.com/wp-graphql/wp-graphql/issues/1024

**Comment by esamattis** [comment]
Actually wp-graphql-polylang cannot work at all when called from PHP. That's because Polylang initializes itself into a specific mode which maybe be "admin", "frontend" or "rest" and now via wp-graphql-polylang "graphl" too. Once the intialization has been done it cannot changed anymore (at least to my knowledge).

So in wp-graphql-polylang I must tell it to go to the graphql mode as soon as possible. If it is too late Polylang will initialize into the frontend mode. This does happen before the init. The frontend mode would filter everything by the default language which not good for a graphql api because you'll want to specify the language using the where-args so you can query posts in multiple languages if you want.

**Comment by jasonbahl** [comment]
@esamattis ugh. Sounds like some funky stuff happening with the Polylang plugin then. 🤔 

I get it though. Most Plugin developers didn't think to plan for a Headless WordPress future when they set out to solve their problems.

**Comment by kidunot89** [comment]
@jasonbahl In WooGraphQL I have one [hook](https://github.com/wp-graphql/wp-graphql-woocommerce/blob/develop/includes/class-woocommerce-filters.php#L32-#L34) in particular that is contingent on the fact that it will only be run during WPGraphQL request, this is the hook for setting the WooCommerce session handler class. If it is used at anytime other than a WPGraphQL request,it will break WooCommerce sessions.

**Comment by jasonbahl** [comment]
@kidunot89 you do that stuff based on custom headers, right?

You probably should do what you need to do with the custom headers to set identity before _any_ processing occurs, not _just_ GraphQL

That's how WordPress in general works. User identity is setup early, then processing occurs based on that identity. 

Identity / Auth is at a layer before processing. You can auth with Basic Auth, Oauth, JWT, or in your case identify yourself with a session token, before processing. 

Then whatever processing occurs, REST, GraphQL, loading a template or admin, should respect that

https://codex.wordpress.org/Plugin_API/Action_Reference

**Comment by jasonbahl** [comment]
closed by #1022 and #1025
